In [2]:
import pandas as pd
import numpy as np
import time

We will analyze the fifa 19 data 

In [3]:
data=pd.read_csv('data.csv')

.head() method is used to display the first 'n' rows of the dataframe. It is like the "top n" command of SQL<br>
Inside the bracket you can specify the number of records to print. Default is 5

In [4]:
data.head(20)

Unnamed: 0      ID               Name  Age  \
0            0  158023           L. Messi   31   
1            1   20801  Cristiano Ronaldo   33   
2            2  190871          Neymar Jr   26   
3            3  193080             De Gea   27   
4            4  192985       K. De Bruyne   27   
5            5  183277          E. Hazard   27   
6            6  177003          L. Modrić   32   
7            7  176580          L. Suárez   31   
8            8  155862       Sergio Ramos   32   
9            9  200389           J. Oblak   25   
10          10  188545     R. Lewandowski   29   
11          11  182521           T. Kroos   28   
12          12  182493           D. Godín   32   
13          13  168542        David Silva   32   
14          14  215914           N. Kanté   27   
15          15  211110          P. Dybala   24   
16          16  202126            H. Kane   24   
17          17  194765       A. Griezmann   27   
18          18  192448      M. ter Stegen   26   
19          19  192119        T. Courtois   26   

                                             Photo Nationality  \
0   https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1    https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2   https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3   https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4   https://cdn.sofifa.org/players/4/19/192985.png     Belgium   
5   https://cdn.sofifa.org/players/4/19/183277.png     Belgium   
6   https://cdn.sofifa.org/players/4/19/177003.png     Croatia   
7   https://cdn.sofifa.org/players/4/19/176580.png     Uruguay   
8   https://cdn.sofifa.org/players/4/19/155862.png       Spain   
9   https://cdn.sofifa.org/players/4/19/200389.png    Slovenia   
10  https://cdn.sofifa.org/players/4/19/188545.png      Poland   
11  https://cdn.sofifa.org/players/4/19/182521.png     Germany   
12  https://cdn.sofifa.org/players/4/19/182493.png     Uruguay   
13  https://cdn.sofifa.org/players/4/19/168542.png       Spain   
14  https://cdn.sofifa.org/players/4/19/215914.png      France   
15  https://cdn.sofifa.org/players/4/19/211110.png   Argentina   
16  https://cdn.sofifa.org/players/4/19/202126.png     England   
17  https://cdn.sofifa.org/players/4/19/194765.png      France   
18  https://cdn.sofifa.org/players/4/19/192448.png     Germany   
19  https://cdn.sofifa.org/players/4/19/192119.png     Belgium   

                                   Flag  Overall  Potential  \
0   https://cdn.sofifa.org/flags/52.png       94         94   
1   https://cdn.sofifa.org/flags/38.png       94         94   
2   https://cdn.sofifa.org/flags/54.png       92         93   
3   https://cdn.sofifa.org/flags/45.png       91         93   
4    https://cdn.sofifa.org/flags/7.png       91         92   
5    https://cdn.sofifa.org/flags/7.png       91         91   
6   https://cdn.sofifa.org/flags/10.png       91         91   
7   https://cdn.sofifa.org/flags/60.png       91         91   
8   https://cdn.sofifa.org/flags/45.png       91         91   
9   https://cdn.sofifa.org/flags/44.png       90         93   
10  https://cdn.sofifa.org/flags/37.png       90         90   
11  https://cdn.sofifa.org/flags/21.png       90         90   
12  https://cdn.sofifa.org/flags/60.png       90         90   
13  https://cdn.sofifa.org/flags/45.png       90         90   
14  https://cdn.sofifa.org/flags/18.png       89         90   
15  https://cdn.sofifa.org/flags/52.png       89         94   
16  https://cdn.sofifa.org/flags/14.png       89         91   
17  https://cdn.sofifa.org/flags/18.png       89         90   
18  https://cdn.sofifa.org/flags/21.png       89         92   
19   https://cdn.sofifa.org/flags/7.png       89         90   

                   Club  ... Composure Marking StandingTackle  SlidingTackle  \
0          FC Barcelona  ...      96.0    33.0           28.0           26.0   
1              Juventus  ...      95.0    28.0           31.0           23.0   
2 

To check the dimensions of the dataframe, we use the method .shape

In [5]:
data.shape

(18207, 89)

We see that the data has 18207 rows and 89 columns <br>

Now lets see how can we get the names of the columns

In [6]:
data.columns

Index(['Unnamed: 0', 'ID', 'Name', 'Age', 'Photo', 'Nationality', 'Flag',
       'Overall', 'Potential', 'Club', 'Club Logo', 'Value', 'Wage', 'Special',
       'Preferred Foot', 'International Reputation', 'Weak Foot',
       'Skill Moves', 'Work Rate', 'Body Type', 'Real Face', 'Position',
       'Jersey Number', 'Joined', 'Loaned From', 'Contract Valid Until',
       'Height', 'Weight', 'LS', 'ST', 'RS', 'LW', 'LF', 'CF', 'RF', 'RW',
       'LAM', 'CAM', 'RAM', 'LM', 'LCM', 'CM', 'RCM', 'RM', 'LWB', 'LDM',
       'CDM', 'RDM', 'RWB', 'LB', 'LCB', 'CB', 'RCB', 'RB', 'Crossing',
       'Finishing', 'HeadingAccuracy', 'ShortPassing', 'Volleys', 'Dribbling',
       'Curve', 'FKAccuracy', 'LongPassing', 'BallControl', 'Acceleration',
       'SprintSpeed', 'Agility', 'Reactions', 'Balance', 'ShotPower',
       'Jumping', 'Stamina', 'Strength', 'LongShots', 'Aggression',
       'Interceptions', 'Positioning', 'Vision', 'Penalties', 'Composure',
       'Marking', 'StandingTackle', 'SlidingT

The method info() is used to print the basic details about the dataframe

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18207 entries, 0 to 18206
Data columns (total 89 columns):
Unnamed: 0                  18207 non-null int64
ID                          18207 non-null int64
Name                        18207 non-null object
Age                         18207 non-null int64
Photo                       18207 non-null object
Nationality                 18207 non-null object
Flag                        18207 non-null object
Overall                     18207 non-null int64
Potential                   18207 non-null int64
Club                        17966 non-null object
Club Logo                   18207 non-null object
Value                       18207 non-null object
Wage                        18207 non-null object
Special                     18207 non-null int64
Preferred Foot              18159 non-null object
International Reputation    18159 non-null float64
Weak Foot                   18159 non-null float64
Skill Moves                 18159 non-null fl

<br>We can change the data type of the column by using a method called .astype()

In [8]:
data['Age']=data['Age'].astype(np.float64)

In [9]:
data['Age'].dtype

dtype('float64')

.describe() method is used to print out the common statistics for numeric columns of the data frame(like mean,median,max,min)

In [10]:
data.describe()

Unnamed: 0             ID           Age       Overall     Potential  \
count  18207.000000   18207.000000  18207.000000  18207.000000  18207.000000   
mean    9103.000000  214298.338606     25.122206     66.238699     71.307299   
std     5256.052511   29965.244204      4.669943      6.908930      6.136496   
min        0.000000      16.000000     16.000000     46.000000     48.000000   
25%     4551.500000  200315.500000     21.000000     62.000000     67.000000   
50%     9103.000000  221759.000000     25.000000     66.000000     71.000000   
75%    13654.500000  236529.500000     28.000000     71.000000     75.000000   
max    18206.000000  246620.000000     45.000000     94.000000     95.000000   

            Special  International Reputation     Weak Foot   Skill Moves  \
count  18207.000000              18159.000000  18159.000000  18159.000000   
mean    1597.809908                  1.113222      2.947299      2.361308   
std      272.586016                  0.394031      0.660456      0.756164   
min      731.000000                  1.000000      1.000000      1.000000   
25%     1457.000000                  1.000000      3.000000      2.000000   
50%     1635.000000                  1.000000      3.000000      2.000000   
75%     1787.000000                  1.000000      3.000000      3.000000   
max     2346.000000                  5.000000      5.000000      5.000000   

       Jersey Number  ...     Penalties     Composure       Marking  \
count   18147.000000  ...  18159.000000  18159.000000  18159.000000   
mean       19.546096  ...     48.548598     58.648274     47.281623   
std        15.947765  ...     15.704053     11.436133     19.904397   
min         1.000000  ...      5.000000      3.000000      3.000000   
25%         8.000000  ...     39.000000     51.000000     30.000000   
50%        17.000000  ...     49.000000     60.000000     53.000000   
75%        26.000000  ...     60.000000     67.000000     64.000000   
max        99.000000  ...     92.000000     96.000000     94.000000   

       StandingTackle  SlidingTackle      GKDiving    GKHandling  \
count    18159.000000   18159.000000  18159.000000  18159.000000   
mean        47.697836      45.661435     16.616223     16.391596   
std         21.664004      21.289135     17.695349     16.906900   
min          2.000000       3.000000      1.000000      1.000000   
25%         27.000000      24.000000      8.000000      8.000000   
50%         55.000000      52.000000     11.000000     11.000000   
75%         66.000000      64.000000     14.000000     14.000000   
max         93.000000      91.000000     90.000000     92.000000   

          GKKicking  GKPositioning    GKReflexes  
count  18159.000000   18159.000000  18159.000000  
mean      16.232061      16.388898     16.710887  
std       16.502864      17.034669     17.955119  
min        1.000000       1.000000      1.000000  
25%        8.000000       8.000000      8.000000  
50%       11.000000      11.000000     11.000000  
75%       14.000000      14.000000     14.000000  
max       91.000000      90.000000     94.000000  

[8 rows x 44 columns]

To get statistics for non numeric objects, we can specify the data type in the describe method

In [11]:
data.describe(include=['object'])

Name                                           Photo  \
count          18207                                           18207   
unique         17194                                           18207   
top     J. Rodríguez  https://cdn.sofifa.org/players/4/19/156481.png   
freq              11                                               1   

       Nationality                                 Flag              Club  \
count        18207                                18207             17966   
unique         164                                  164               651   
top        England  https://cdn.sofifa.org/flags/14.png  Newcastle United   
freq          1662                                 1662                33   

                                           Club Logo  Value   Wage  \
count                                          18207  18207  18207   
unique                                           679    217    144   
top     https://cdn.sofifa.org/teams/2/light/240.png  €1.1M    €1K   
freq                                              33    431   4900   

       Preferred Foot       Work Rate  ...    LDM    CDM    RDM    RWB     LB  \
count           18159           18159  ...  16122  16122  16122  16122  16122   
unique              2               9  ...     99     99     99     95     98   
top             Right  Medium/ Medium  ...   62+2   62+2   62+2   61+2   61+2   
freq            13948            9810  ...    660    660    660    713    669   

          LCB     CB    RCB     RB Release Clause  
count   16122  16122  16122  16122          16643  
unique    108    108    108     98           1244  
top      63+2   63+2   63+2   61+2          €1.1M  
freq      634    634    634    669            557  

[4 rows x 45 columns]

<br>.value_counts() method is used to get the counts of different values for a categorical variable

In [12]:
data['Club'].value_counts()

Newcastle United             33
Empoli                       33
Wolverhampton Wanderers      33
Fortuna Düsseldorf           33
Arsenal                      33
AS Monaco                    33
Borussia Dortmund            33
CD Leganés                   33
FC Barcelona                 33
Tottenham Hotspur            33
Cardiff City                 33
Valencia CF                  33
RC Celta                     33
Everton                      33
Atlético Madrid              33
Real Madrid                  33
Manchester City              33
Southampton                  33
TSG 1899 Hoffenheim          33
Eintracht Frankfurt          33
Rayo Vallecano               33
Frosinone                    33
Liverpool                    33
Manchester United            33
Chelsea                      33
Burnley                      33
Bournemouth                  32
Fulham                       32
Leicester City               32
Huddersfield Town            32
                             ..
Colorado

In [13]:
data['Jersey Number'].value_counts(normalize=True)

8.0     0.033725
7.0     0.033284
10.0    0.032678
11.0    0.032512
6.0     0.032292
5.0     0.031906
9.0     0.031796
4.0     0.031575
20.0    0.031300
1.0     0.031190
17.0    0.030528
3.0     0.030143
23.0    0.030088
19.0    0.030033
18.0    0.030033
14.0    0.029867
21.0    0.029537
22.0    0.029261
2.0     0.028600
16.0    0.028490
15.0    0.027608
24.0    0.023420
27.0    0.023310
13.0    0.023089
25.0    0.022538
12.0    0.021491
26.0    0.021491
30.0    0.020444
29.0    0.019728
28.0    0.019673
          ...   
94.0    0.000606
92.0    0.000551
71.0    0.000551
52.0    0.000551
60.0    0.000551
57.0    0.000441
61.0    0.000386
51.0    0.000386
72.0    0.000386
73.0    0.000331
83.0    0.000331
69.0    0.000331
75.0    0.000331
62.0    0.000331
67.0    0.000276
81.0    0.000276
58.0    0.000276
78.0    0.000276
59.0    0.000276
82.0    0.000276
63.0    0.000220
86.0    0.000220
68.0    0.000220
76.0    0.000220
65.0    0.000220
84.0    0.000220
64.0    0.000110
79.0    0.0001

<br> To sort the dataframe, we use .sort_values(). We can specify the list of columns on which we wish to sort the dataframe

In [14]:
data.sort_values(by='ShotPower', ascending=False).head()

Unnamed: 0      ID               Name   Age  \
1              1   20801  Cristiano Ronaldo  33.0   
378          378  189362               Hulk  31.0   
1348        1348  169051          F. Guarín  32.0   
890          890  150516        L. Podolski  33.0   
36            36  173731            G. Bale  28.0   

                                               Photo Nationality  \
1      https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
378   https://cdn.sofifa.org/players/4/19/189362.png      Brazil   
1348  https://cdn.sofifa.org/players/4/19/169051.png    Colombia   
890   https://cdn.sofifa.org/players/4/19/150516.png     Germany   
36    https://cdn.sofifa.org/players/4/19/173731.png       Wales   

                                     Flag  Overall  Potential  \
1     https://cdn.sofifa.org/flags/38.png       94         94   
378   https://cdn.sofifa.org/flags/54.png       81         81   
1348  https://cdn.sofifa.org/flags/56.png       76         76   
890   https://cdn.sofifa.org/flags/21.png       78         78   
36    https://cdn.sofifa.org/flags/50.png       88         88   

                               Club  ... Composure Marking StandingTackle  \
1                          Juventus  ...      95.0    28.0           31.0   
378                Shanghai SIPG FC  ...      75.0    50.0           41.0   
1348  Shanghai Greenland Shenhua FC  ...      76.0    60.0           66.0   
890                     Vissel Kobe  ...      79.0    54.0           39.0   
36                      Real Madrid  ...      86.0    54.0           55.0   

      SlidingTackle GKDiving  GKHandling  GKKicking  GKPositioning GKReflexes  \
1              23.0      7.0        11.0       15.0           14.0       11.0   
378            28.0      6.0        10.0        7.0           11.0       12.0   
1348           60.0     12.0        13.0       13.0           15.0       13.0   
890            39.0     10.0         9.0       16.0            9.0        9.0   
36             52.0     15.0        15.0       11.0            5.0        6.0   

     Release Clause  
1           €127.1M  
378          €34.1M  
1348         €12.9M  
890           €8.1M  
36            €123M  

[5 rows x 89 columns]

### Indexing and Retrieving

A DataFrame can be indexed in a few different ways.

To get a single column, you can use a DataFrame['Name'] construction. Let's use this to answer a question about that column alone: what is the mean overall rating of all the players in our dataframe?

In [15]:
data['Overall'].mean()

66.23869940132916

<b>Boolean indexing</b> with one column is also very convenient. The syntax is df[P(df['Name'])], where P is some logical condition that is checked for each element of the Name column. The result of such indexing is the DataFrame consisting only of rows that satisfy the P condition on the Name column.

Let’s use it to answer the question:<br>
What is the average rating for Club Manchester United?

In [16]:
data[data.Club=='Manchester United']['Overall'].mean()

77.24242424242425

What is the average rating for players from Spain in Manchester United?

In [17]:
data[(data.Club=='Manchester United') & (data.Nationality=='Spain')]['Overall'].mean()

85.0

DataFrames can be indexed by column name (label) or row name (index) or by the serial number of a row. The loc method is used for indexing by name, while iloc() is used for indexing by number.

In the first case below, we say "give us the values of the rows with index from 0 to 5 (inclusive) and columns labeled from Name to Potential (inclusive)". In the second case, we say "give us the values of the first five rows in the first three columns" (as in a typical Python slice: the maximal value is not included).

In [18]:
data.loc[0:5, 'Name':'Potential']

Name   Age                                           Photo  \
0           L. Messi  31.0  https://cdn.sofifa.org/players/4/19/158023.png   
1  Cristiano Ronaldo  33.0   https://cdn.sofifa.org/players/4/19/20801.png   
2          Neymar Jr  26.0  https://cdn.sofifa.org/players/4/19/190871.png   
3             De Gea  27.0  https://cdn.sofifa.org/players/4/19/193080.png   
4       K. De Bruyne  27.0  https://cdn.sofifa.org/players/4/19/192985.png   
5          E. Hazard  27.0  https://cdn.sofifa.org/players/4/19/183277.png   

  Nationality                                 Flag  Overall  Potential  
0   Argentina  https://cdn.sofifa.org/flags/52.png       94         94  
1    Portugal  https://cdn.sofifa.org/flags/38.png       94         94  
2      Brazil  https://cdn.sofifa.org/flags/54.png       92         93  
3       Spain  https://cdn.sofifa.org/flags/45.png       91         93  
4     Belgium   https://cdn.sofifa.org/flags/7.png       91         92  
5     Belgium   https://cdn.sofifa.org/flags/7.png       91         91

In [19]:
data.iloc[0:5,0:3]

Unnamed: 0      ID               Name
0           0  158023           L. Messi
1           1   20801  Cristiano Ronaldo
2           2  190871          Neymar Jr
3           3  193080             De Gea
4           4  192985       K. De Bruyne

If we need the first or the last line of the data frame, we can use the df[:1] or df[-1:] construct:

In [20]:
data[:1]

Unnamed: 0      ID      Name   Age  \
0           0  158023  L. Messi  31.0   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   

                                  Flag  Overall  Potential          Club  ...  \
0  https://cdn.sofifa.org/flags/52.png       94         94  FC Barcelona  ...   

  Composure Marking StandingTackle  SlidingTackle GKDiving  GKHandling  \
0      96.0    33.0           28.0           26.0      6.0        11.0   

   GKKicking  GKPositioning GKReflexes Release Clause  
0       15.0           14.0        8.0        €226.5M  

[1 rows x 89 columns]

### Applying functions to cells, rows and columns

#### To apply functions to each column, use apply():

In [21]:
data.loc[:,'Crossing':'GKReflexes'].apply(np.max)

Crossing           93.0
Finishing          95.0
HeadingAccuracy    94.0
ShortPassing       93.0
Volleys            90.0
Dribbling          97.0
Curve              94.0
FKAccuracy         94.0
LongPassing        93.0
BallControl        96.0
Acceleration       97.0
SprintSpeed        96.0
Agility            96.0
Reactions          96.0
Balance            96.0
ShotPower          95.0
Jumping            95.0
Stamina            96.0
Strength           97.0
LongShots          94.0
Aggression         95.0
Interceptions      92.0
Positioning        95.0
Vision             94.0
Penalties          92.0
Composure          96.0
Marking            94.0
StandingTackle     93.0
SlidingTackle      91.0
GKDiving           90.0
GKHandling         92.0
GKKicking          91.0
GKPositioning      90.0
GKReflexes         94.0
dtype: float64

The apply method can also be used to apply a function to each row. To do this, specify axis=1. Lambda functions are very convenient in such scenarios.
<br>For example, if we need to select all players whose name starts with 'M', we can do it like this:

In [22]:
data[data['Name'].apply(lambda Name: Name[0] == 'M')].head()

Unnamed: 0      ID           Name   Age  \
18          18  192448  M. ter Stegen  26.0   
22          22  167495       M. Neuer  32.0   
26          26  209331       M. Salah  26.0   
34          34  178603     M. Hummels  29.0   
35          35  176676        Marcelo  30.0   

                                             Photo Nationality  \
18  https://cdn.sofifa.org/players/4/19/192448.png     Germany   
22  https://cdn.sofifa.org/players/4/19/167495.png     Germany   
26  https://cdn.sofifa.org/players/4/19/209331.png       Egypt   
34  https://cdn.sofifa.org/players/4/19/178603.png     Germany   
35  https://cdn.sofifa.org/players/4/19/176676.png      Brazil   

                                    Flag  Overall  Potential  \
18   https://cdn.sofifa.org/flags/21.png       89         92   
22   https://cdn.sofifa.org/flags/21.png       89         89   
26  https://cdn.sofifa.org/flags/111.png       88         89   
34   https://cdn.sofifa.org/flags/21.png       88         88   
35   https://cdn.sofifa.org/flags/54.png       88         88   

                 Club  ... Composure Marking StandingTackle  SlidingTackle  \
18       FC Barcelona  ...      69.0    25.0           13.0           10.0   
22  FC Bayern München  ...      70.0    17.0           10.0           11.0   
26          Liverpool  ...      91.0    38.0           43.0           41.0   
34  FC Bayern München  ...      91.0    88.0           90.0           88.0   
35        Real Madrid  ...      86.0    71.0           85.0           86.0   

   GKDiving  GKHandling  GKKicking  GKPositioning GKReflexes Release Clause  
18     87.0        85.0       88.0           85.0       90.0        €123.3M  
22     90.0        86.0       91.0           87.0       87.0         €62.7M  
26     14.0        14.0        9.0           11.0       14.0        €137.3M  
34     15.0         6.0       10.0            5.0        6.0         €75.9M  
35     12.0         5.0        5.0            5.0        9.0         €88.2M  

[5 rows x 89 columns]

### Grouping<br>
In general, grouping data in Pandas works as follows:

df.groupby(by=grouping_columns)[columns_to_show].function()<br>
<li>First, the groupby method divides the grouping_columns by their values. They become a new index in the resulting dataframe.</li>
<li>Then, columns of interest are selected (columns_to_show). If columns_to_show is not included, all non groupby clauses will be included.</li>
<li>Finally, one or several functions are applied to the obtained groups per selected columns.</li> <br>

Here's an example of getting club wise avg overall rating

In [27]:
data.groupby(['Club'])['Overall'].mean().sort_values(ascending=False).head(10)

Club
Juventus               82.280000
Napoli                 80.000000
Inter                  79.750000
Real Madrid            78.242424
Milan                  78.074074
FC Barcelona           78.030303
Paris Saint-Germain    77.433333
Roma                   77.423077
Manchester United      77.242424
SL Benfica             77.000000
Name: Overall, dtype: float64

Let’s do the same thing, but slightly differently by passing a list of functions to agg():

In [34]:
data.groupby(['Club'])['Overall'].agg([np.mean,np.max,np.min]).head(10)

mean  amax  amin
Club                                        
 SSV Jahn Regensburg   65.586207    71    52
1. FC Heidenheim 1846  65.750000    75    53
1. FC Kaiserslautern   63.384615    70    54
1. FC Köln             70.785714    83    53
1. FC Magdeburg        65.615385    71    56
1. FC Nürnberg         68.827586    76    59
1. FC Union Berlin     68.321429    73    53
1. FSV Mainz 05        70.843750    79    58
AC Ajaccio             65.043478    73    53
AC Horsens             60.640000    68    50

### Summary tables
Suppose we want to see how the observations in our sample are distributed in the context of two variables - Churn and International plan. To do so, we can build a contingency table using the crosstab method:

In [35]:
pd.crosstab(data['Nationality'], data['Club'])

Club                   SSV Jahn Regensburg  1. FC Heidenheim 1846  \
Nationality                                                         
Afghanistan                              0                      0   
Albania                                  0                      0   
Algeria                                  0                      0   
Andorra                                  0                      0   
Angola                                   0                      0   
Antigua & Barbuda                        0                      0   
Argentina                                0                      0   
Armenia                                  1                      0   
Australia                                0                      0   
Austria                                  0                      1   
Azerbaijan                               0                      0   
Barbados                                 0                      0   
Belarus                                  0                      0   
Belgium                                  0                      0   
Belize                                   0                      0   
Benin                                    0                      0   
Bermuda                                  0                      0   
Bolivia                                  0                      0   
Bosnia Herzegovina                       0                      0   
Botswana                                 0                      0   
Brazil                                   0                      0   
Bulgaria                                 0                      0   
Burkina Faso                             0                      0   
Burundi                                  0                      0   
Cameroon                                 0                      0   
Canada                                   0                      0   
Cape Verde                               0                      0   
Central African Rep.                     0                      0   
Chad                                     0                      0   
Chile                                    0                      0   
...                                    ...                    ...   
Sierra Leone                             0                      0   
Slovakia                                 0                      0   
Slovenia                                 0                      0   
South Africa                             0                      0   
South Sudan                              0                      0   
Spain                                    0                      0   
St Kitts Nevis                           0                      0   
St Lucia                                 0                      0   
Sudan                                    0                      0   
Suriname                                 0                      0   
Sweden                                   0                      0   
Switzerland                              0                      0   
Syria                                    0                      0   
São Tomé & Príncipe                      0                      0   
Tanzania                                 0                      0   
Thailand                                 0                      0   
Togo                                     0                      0   
Trinidad & Tobago                        0                      0   
Tunisia                                  0                      0   
Turkey                                   0                      0   
Uganda                                   0                      0   
Ukraine                                  0                      0   
United Arab Emirates                     0                      0   
United States                            0                      0   
Uruguay                                  0                      0  

Pivot tables are implemented in Pandas: the pivot_table method takes the following parameters:

values – a list of variables to calculate statistics for,<br>
index – a list of variables to group data by,<br>
aggfunc – what statistics we need to calculate for groups, ex. sum, mean, maximum, minimum or something else.<br>

Let's look at average of sprint speed, shooting, strength

In [40]:
data.pivot_table(['SprintSpeed', 'ShotPower', 'Strength'],
               ['Club'], aggfunc='mean').head(10)

ShotPower  SprintSpeed   Strength
Club                                                    
 SSV Jahn Regensburg   54.137931    61.034483  69.758621
1. FC Heidenheim 1846  57.071429    64.678571  66.214286
1. FC Kaiserslautern   56.692308    64.692308  66.576923
1. FC Köln             58.178571    64.250000  66.428571
1. FC Magdeburg        52.269231    64.269231  70.230769
1. FC Nürnberg         55.103448    65.344828  65.344828
1. FC Union Berlin     58.964286    65.714286  68.750000
1. FSV Mainz 05        58.187500    69.187500  68.156250
AC Ajaccio             54.565217    64.565217  64.347826
AC Horsens             53.960000    59.960000  68.160000

### DataFrame transformations
Like many other things in Pandas, adding columns to a DataFrame is doable in many ways.

For example, if we want to add a column which gives the difference in the overall and potential of a player, we can create it directly 

In [45]:
data['gap_to_potential']=data['Potential']-data['Overall']
data.sort_values(by='gap_to_potential',ascending=False).head(10)

Unnamed: 0      ID           Name   Age  \
18072       18072  238758    D. Campbell  17.0   
16028       16028  244592    J. von Moos  17.0   
17175       17175  243353       B. Mumba  16.0   
16629       16629  246241       Y. Lenze  17.0   
17784       17784  246100   K. Askildsen  17.0   
13927       13927  241403        A. Dabo  17.0   
18050       18050  245370  E. Destanoglu  17.0   
18166       18166  243621       N. Ayéva  16.0   
15728       15728  245721   G. Azzinnari  17.0   
18044       18044  246109        K. Lara  16.0   

                                                Photo  Nationality  \
18072  https://cdn.sofifa.org/players/4/19/238758.png     Scotland   
16028  https://cdn.sofifa.org/players/4/19/244592.png  Switzerland   
17175  https://cdn.sofifa.org/players/4/19/243353.png      England   
16629  https://cdn.sofifa.org/players/4/19/246241.png      Germany   
17784  https://cdn.sofifa.org/players/4/19/246100.png       Norway   
13927  https://cdn.sofifa.org/players/4/19/241403.png       France   
18050  https://cdn.sofifa.org/players/4/19/245370.png       Turkey   
18166  https://cdn.sofifa.org/players/4/19/243621.png       Sweden   
15728  https://cdn.sofifa.org/players/4/19/245721.png        Italy   
18044  https://cdn.sofifa.org/players/4/19/246109.png     Colombia   

                                      Flag  Overall  Potential  \
18072  https://cdn.sofifa.org/flags/42.png       50         76   
16028  https://cdn.sofifa.org/flags/47.png       58         84   
17175  https://cdn.sofifa.org/flags/14.png       55         80   
16629  https://cdn.sofifa.org/flags/21.png       57         82   
17784  https://cdn.sofifa.org/flags/36.png       52         77   
13927  https://cdn.sofifa.org/flags/18.png       62         86   
18050  https://cdn.sofifa.org/flags/48.png       50         74   
18166  https://cdn.sofifa.org/flags/46.png       48         72   
15728  https://cdn.sofifa.org/flags/27.png       59         83   
18044  https://cdn.sofifa.org/flags/56.png       50         74   

                 Club  ... Marking StandingTackle SlidingTackle  GKDiving  \
18072        Aberdeen  ...    32.0           40.0          24.0      11.0   
16028   FC Basel 1893  ...    36.0           16.0          12.0      11.0   
17175      Sunderland  ...    55.0           56.0          54.0      14.0   
16629   FC Schalke 04  ...    11.0           14.0          13.0      58.0   
17784  Stabæk Fotball  ...    43.0           45.0          51.0       9.0   
13927       FC Nantes  ...    62.0           65.0          63.0       8.0   
18050     Beşiktaş JK  ...     6.0           13.0          12.0      52.0   
18166       Örebro SK  ...    14.0           19.0          16.0      10.0   
15728         Cosenza  ...    42.0           25.0          20.0      14.0   
18044  Atlético Huila  ...    47.0           50.0          47.0       9.0   

      GKHandling  GKKicking  GKPositioning  GKReflexes Release Clause  \
18072       14.0        9.0           14.0        11.0          €165K   
16028       13.0       12.0           14.0         8.0          €700K   
17175        9.0        8.0           11.0        14.0          €523K   
16629       55.0       63.0           51.0        60.0          €513K   
17784       12.0        7.0           10.0         8.0          €248K   
13927        8.0       14.0            5.0        10.0          €1.9M   
18050       52.0       50.0           47.0        52.0          €145K   
18166        9.0        8.0            7.0         6.0          €158K   
15728       12.0       13.0           10.0        10.0          €893K   
18044       13.0        8.0           12.0         9.0          €141K   

      gap_to_potential  
18072               26  
16028               26  
17175               25  
16629               25  
17784               25  
13927               24  
18050               24  
18166               24  
15728               24  
18044               24  

[10 rows x 90 columns]

To delete columns or rows, use the drop method, passing the required indexes and the axis parameter (1 if you delete columns, and nothing or 0 if you delete rows). The inplace argument tells whether to change the original DataFrame. With inplace=False, the drop method doesn't change the existing DataFrame and returns a new one with dropped rows or columns. With inplace=True, it alters the DataFrame.

In [47]:
data.drop(['Photo', 'Flag'], axis=1, inplace=True)
data.head(10)

KeyError: "['Photo' 'Flag'] not found in axis"

In [ ]:
# and here’s how you can delete rows
data.drop([1, 2]).head() 